File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # window = [0]*26
        # given = [0] *26

        # if len(s1) > len(s2):
        #     return False

        # for i in range (len(s1)):
        #     given[ord(s1[i])-ord('a')]+=1

        # for i in range (len(s1)):
        #     window[ord(s2[i])-ord('a')]+=1

        # if given == window:
        #     return True

        # for i in range(len(s1),len(s2)):
        #     window[ord(s2[i])-ord('a')]+=1
        #     window[ord(s2[i-len(s1)])-ord('a')]-=1
        #     if given == window:
        #         return True
        # return False
        giv = {}
        ans = {}
        count = 0
        for i in range(len(s1)):
            giv[s1[i]] = giv.get(s1[i], 0) + 1
        l = 0
        for r in range(len(s2)):
            ans[s2[r]] = ans.get(s2[r], 0) + 1

            if r - l + 1 > len(s1):
                ans[s2[l]] -= 1
                if ans[s2[l]] == 0:
                    del ans[s2[l]]
                l += 1
            if giv == ans:
                return True
        return False

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length()>s2.length()) return false;
        int formed = 0 , required =0 , l = 0;
        HashMap<Character,Integer> mp = new HashMap<>();
        HashMap<Character,Integer> window = new HashMap<>();
        for(int i=0;i<s1.length();i++){
            char temp = s1.charAt(i);
            mp.put(temp,mp.getOrDefault(temp,0)+1);
        }
        required = mp.size();
        for(int r=0;r<s2.length();r++){
            char temp = s2.charAt(r);
            window.put(temp,window.getOrDefault(temp,0)+1);
            if(mp.containsKey(temp) && window.get(temp).intValue() == mp.get(temp).intValue()){
                formed ++;
            }
            while(r-l+1 > s1.length()){
                if(mp.containsKey(s2.charAt(l)) && window.get(s2.charAt(l)).intValue() == mp.get(s2.charAt(l)).intValue()){
                    formed--;
                }
                window.put(s2.charAt(l),window.get(s2.charAt(l))-1);
                l++;
            }
            if(formed == required  ) return true;
        }
        return false;
    }
}